iroh-net/src/discovery/local_swarm_discovery.rs

```diff
@@ -81,6 +86,36 @@ enum Message {
     Subscribe(mpsc::Sender<(NodeId, DiscoveryItem)>),
 }
 
+/// Manages the list of subscribers that are subscribed to this discovery service.
+struct Subscribers(Vec<mpsc::Sender<(NodeId, DiscoveryItem)>>);
+
+impl Subscribers {
+    fn new() -> Self {
+        Self(vec![])
+    }
+
+    /// Add the subscriber to the list of subscribers
+    fn push(&mut self, subscriber: mpsc::Sender<(NodeId, DiscoveryItem)>) {
+        self.0.push(subscriber);
+    }
+
+    /// Sends the `node_id` and `item` to each subscriber.
+    ///
+    /// Cleans up any subscribers that have been dropped.
+    async fn send(&mut self, node_id: NodeId, item: DiscoveryItem) {
+        let mut clean_up = vec![];
+        for (i, subscriber) in self.0.iter().enumerate() {
+            // assume subscriber was dropped
+            if (subscriber.send((node_id, item.clone())).await).is_err() {
+                clean_up.push(i);
+            }
+        }
+        for i in clean_up {
+            self.0.swap_remove(i);
+        }
```

Comment:
this will panic if any of the removed elements is among the last ones and more than their index is removed.

[playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021)

Reply:
I think that for this to work you would need to iterate the indexes in reverse

Reply:
nice catch, thank you.

reversing will work. in the worse case scenario, where we remove every single element in reverse order, every element will just pop off the end of the list. ty ty